packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -41,16 +42,38 @@ const isLiteral = (type: ts.Type): boolean =>
   isLiteralType(type);
 // #endregion
 
+// Array type utilities
+// #region
+function isTypeReference(type: ts.Type): type is ts.TypeReference {
+  return !!(
+    type.getFlags() & ts.TypeFlags.Object &&
+    (type as ts.ObjectType).objectFlags & ts.ObjectFlags.Reference
+  );
+}
+
+// There's a built-in checker.isArrayType, but it's marked as internal.
+function isArrayType(type: ts.Type): boolean {
+  return (
+    isTypeReference(type) &&
+    (type.target.symbol.name === 'Array' ||
+      type.target.symbol.name === 'ReadonlyArray')
+  );
+}
+// #endregion
```

Comment:
Feel free to add this file into your PR, and use the API.
https://github.com/typescript-eslint/typescript-eslint/blob/naming-conventions/packages/eslint-plugin/typings/typescript.d.ts

The API is "internal", but it's stable - they just haven't gotten around to exposing their type relationship API yet (there's a whole discussion around why they haven't yet...)